_posts/2020-11-01-batch-processing-stop-job-step.adoc

```diff
@@ -0,0 +1,229 @@
+---
+layout: post
+title: How to properly stop batch processing job and step in WildFly
+date: 2020-11-01
+tags: wildfly jberet batch
+synopsis: How to properly stop a running batch job with API and WildFly tools, and how to stop an individual step.
+author: chengfang
+---
+
+Batch jobs are long-running background processing tasks, and therefore it's
+common that user may need to pause or stop the execution.  WildFly, which
+implements batch processing through its jberet component based on JSR 352
+and https://github.com/eclipse-ee4j/batch-api[Jakarta Batch Specification],
+offers a variety of ways to achieve that requirement.  This post will
+demonstrate how to stop a running job execution, or individual step execution,
+and some design and implementation consideration.
+
+## Properly stop a running job execution
+
+Batch spec defines a standard API for stopping a running job execution:
+https://github.com/eclipse-ee4j/batch-api/blob/master/api/src/main/java/jakarta/batch/operations/JobOperator.java#L198[javax.batch.operations.JobOperator#stop]
+
+As stated in its https://javaee.github.io/javaee-spec/javadocs/javax/batch/operations/JobOperator.html#stop-long-[javadoc],
+calling `JobOperator.stop()` sends a stop request to the batch container, which tries its best effort to stop
+the running job execution. So it's important to implement batch application that properly responds to a stop request.
+In the following sections, I'll explain what it entails for the two different types of steps: batchlet step and chunk step.
+
+### Stop a running job execution that contains batchlet step
+
+A batchelt step represents a free-form, opaque task that is fully controlled by the batch application. The batch container
+has no chance to intervene once the batchlet starts its processing. So the batchlet class is responsible for providing
+a way to stop itself, if it wants to support graceful stop. That's why `javax.batch.api.Batchlet` interface declares a
+`stop()` method that a batchlet class must implement.
+
+In the example batchlet class below, once it receives a stop request in its `stop()` method, it sets the `toStop` flag
+to `true`. Its `process()` method periodically checks this flag to determine if it needs to stop processing.
+
+Note that bathlet `stop()` method is called asynchronously while the batchlet `process()` method is running in a
+different thread. And the batchlet class should be implemented to properly handle concurrency.
+
+[source,java]
+----
+@Named
+public class Batchlet1 implements Batchlet {
+    private final AtomicBoolean toStop = new AtomicBoolean();
+
+    @Override
+    public String process() throws Exception {
+        String exitStatus = "BATCHLET1_COMPLETED";
+        while (true) {
+            if (toStop.get()) {
+                exitStatus = "BATCHLET1_STOPPED";
+                break;
+            }
+            // perform batchlet task
+            // Thread.sleep(5000);
```

Comment:
Should this be commented out?

Reply:
looks like the sleep line can be confusing. So I removed this sleep line and expand the //perform batchlet task to include some examples.